### ISSUE TITLE
RetryHandler with v6.5.1 coming from v5.80.0

### ISSUE THREAD

**Issue by mambord** [microsoftgraph/msgraph-sdk-java#1933]
I am currently working on a task to upgrade our application from msgraph-sdk-java 5.80.0 to 6.5.1. We have in this application one Graph class to provide all needed functions, like getting applications, service principals, settings owners, etc.

For our use case, we need a RetryHandler if the return code of a request is 404. We achieved this with version 5.x by implementing a RetryHandler and configuring it with the graph client instantiation.

With the new version 6.x, the approach is to configure this per request by passing a RetryHandlerOption to the default RetryHandler Middleware. This leads me to the first question:

**Is it indeed not an option anymore, to configure a RetryHandler "globally" at client instantiation?** In our Graph class, we have tons of requests - and it seems quite a pain to add the RetryHandlerOption to every one of these. I looked at the default RetryHandler Middleware and thought about taking the same approach as before by implementing our own RetryHandler. But there are package-private classes in use and simply copying from the default one, seems not to be possible.

And now to my second question: When I configure a RetryHandlerOption I get a `java.lang.UnsupportedOperationException`. I tried to begin with a minimal setup:

The client class (client initialisation & test function):
```java

    private final val clientSecretCredential: ClientSecretCredential? =
        ClientSecretCredentialBuilder()
            .clientId(azureConfig.graphApiClientId)
            .clientSecret(azureConfig.graphApiClientSecret)
            .tenantId(azureConfig.tenantId)
            .build()

    private val authProvider = AzureIdentityAuthenticationProvider(clientSecretCredential, azureConfig.allowedHosts, azureConfig.scopeGraphApi)
    private final val msGraphApiDebugger = MsGraphApiDebugger()
    private final val retryHandler = RetryHandler()

    private final val httpClient = GraphClientFactory.create()
        .addInterceptor(msGraphApiDebugger)
        .addInterceptor(retryHandler)
        .build()

    fun getApplicationByObjectId(objectId: String): Application {
        val retryHandlerOption = RetryHandlerOption()

        return graphServiceClient.applications().byApplicationId(oid).get { requestConfiguration ->
            requestConfiguration.options.add(retryHandlerOption)
        }
    }
```

The unit test:
```
    @Test
    fun `check that getting an existing app registration returns the app registration`() {
        val application = graphApiClient.getApplicationByObjectId("ad86b88a-bd06-4299-a587-1da6c3d0abdd")
        assertEquals("Example App Registration Display Name", application.displayName)
    }
```

The exception:
```
java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractList.add(AbstractList.java:155)
	at java.base/java.util.AbstractList.add(AbstractList.java:113)
	at xx.xxx.xxx.client.MsGraphApiClient.getApplicationByObjectId$lambda$0(MsGraphApiClient.kt:60)
	at com.microsoft.kiota.RequestInformation.configure(RequestInformation.java:81)
	at com.microsoft.graph.applications.item.ApplicationItemRequestBuilder.toGetRequestInformation(ApplicationItemRequestBuilder.java:335)
	at com.microsoft.graph.applications.item.ApplicationItemRequestBuilder.get(ApplicationItemRequestBuilder.java:267)
	at xx.xxx.xxx.client.MsGraphApiClient.getApplicationByObjectId(MsGraphApiClient.kt:59)
	at xx.xxx.xxx.client.MsGraphApiClientIT.check that getting an existing app registration returns the app registration(MsGraphApiClientIT.kt:16)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

**Do you have an idea what I am doing wrong?** Let me know if you need further informations...

Regards

**Comment by baywet** [comment]
Hi @mambord 
Thanks for using the Java SDK and for reaching out.
The issue is specifically caused by the lambda body:
`requestConfiguration.options.add(retryHandlerOption)`

Which leads me to ask this question: what type does `RetryHandlerOption` return? Where does it come from?

Additionally your snippet doesn't include the creation of the graph client from the http client and authentication provider, could you share that as well please?

Besides those two aspects, the way you're injecting a new middleware handler (interceptor) is the correct approach for cross cutting concerns you don't want to repeat.

**Comment by mambord** [comment]
Hi @baywet 

> Which leads me to ask this question: what type does RetryHandlerOption return? Where does it come from?

RetryHandlerOption does instantiate RetryHandlerOption Class from "com.microsoft.kiota.http.middleware.options" with default settings. To go one step further, my approach to handle retries for a 404 would be like:

```
fun getApplicationByObjectId(objectId: String): Application {
        val iShouldRetry = IShouldRetry { _, _, _, response ->
            response.code == 404
        }
        
        val retryHandlerOption = RetryHandlerOption(iShouldRetry, 3, 3L)

        return graphServiceClient.applications().byApplicationId(objectId).get { requestConfiguration ->
            requestConfiguration.options.add(retryHandlerOption)
        }
    }
```

> Additionally your snippet doesn't include the creation of the graph client from the http client and authentication provider, could you share that as well please?

```
    private final val clientSecretCredential: ClientSecretCredential? =
        ClientSecretCredentialBuilder()
            .clientId(azureConfig.graphApiClientId)
            .clientSecret(azureConfig.graphApiClientSecret)
            .tenantId(azureConfig.tenantId)
            .build()

    private val authProvider = AzureIdentityAuthenticationProvider(
        clientSecretCredential,
        azureConfig.allowedHosts,
        azureConfig.scopeGraphApi
    )

    private final val msGraphApiDebugger = MsGraphApiDebugger()
    private final val retryHandler = RetryHandler()

    private final val httpClient = GraphClientFactory.create()
        .addInterceptor(msGraphApiDebugger)
        .addInterceptor(retryHandler)
        .build()

    private val graphServiceClient = GraphServiceClient(authProvider, httpClient)
```

**Comment by baywet** [comment]
Besides the fact that you're adding the handler a second time when it's already present, everything looks good to me.
Maybe you have multiple versions of the kiota assemblies?

**Comment by m-moris** [comment]
I encountered the same problem as the second one. When I add a handler as follows, `java.lang.UnsupportedOperationException` occurs.  
   
```java  
        var user = graphClient.users().byUserId(id).get(requestConfiguration -> {  
            requestConfiguration.options.add(new RetryHandlerOption());  
        });  
```  
   
To simplify the discussion, let's stop the lambda expression.  
   
```java  
        Consumer<com.microsoft.graph.users.item.UserItemRequestBuilder.GetRequestConfiguration> requestConfiguration =  
                new Consumer<>() {  
                    @Override  
                    public void accept(UserItemRequestBuilder.GetRequestConfiguration requestConfiguration) {  
                        requestConfiguration.options.add(new RetryHandlerOption());  
                    }  
                };  
  
        var user = graphClient.users().byUserId(id).get(requestConfiguration);  
```  
   
You can see that `UserItemRequestBuilder.GetRequestConfiguration#options` is being used.  
   
`GetRequestConfiguration` inherits `com.microsoft.kiota.BaseRequestConfiguration`.  
And the `options` field is assigned `Collections.emptyList()`, which is immutable, thus an exception occurs.  

https://github.com/microsoft/kiota-java/blob/5079c15eb16e2ce363585db1a19e83b6244d5ff7/components/abstractions/src/main/java/com/microsoft/kiota/BaseRequestConfiguration.java#L27

In conclusion, I think the Microsoft Graph Java SDK v6 Changelog and Upgrade Guide documentation is incorrect.  

https://github.com/microsoftgraph/msgraph-sdk-java/blob/dev/docs/upgrade-to-v6.md#option-class-removal
  
You can avoid the problem by assigning as follows.  
   
```java  
        var user = graphClient.users().byUserId(id).get(requestConfiguration -> {  
            requestConfiguration.options = Arrays.asList(new RetryHandlerOption());  
        });  
```

**Comment by m-moris** [comment]
I created an Issue just in case, and there is no problem to Pull Request it.
